### PR TITLE
WAR on index mapping when exact and permissive maps differ

### PR DESCRIPTION
**Pull request by shmsong**
This is a WAR for [this upstream issue](https://github.com/pytorch/pytorch/issues/84510):

Very counter-intuitively, in ca maps, exact mapping does not imply permissive mapping. The war included in this PR should be a safe change but it isn't a thorough fix of this issue.

The underlying issue is quite similar to [this one](https://github.com/csarofeen/pytorch/pull/1930#issuecomment-1232175340) even though they ended up hitting different asserts.

As shown in the printout from the repro in this PR:
```

%kernel {
T2_l[ bS17{( 1 * 1 )} ]
   = T0_g[ bS18{( 1 * 1 )} ];
T3_l[ bS16{( 1 * ( 1 * 1 ) )} ]
   = broadcast( T2_l[ bS17{( 1 * 1 )} ] )
T4_g[ iS14{( i4 * ( 1 * 1 ) )} ]
   = T3_l[ bS16{( 1 * ( 1 * 1 ) )} ]
   + T1_g[ iS20{( i4 * ( 1 * 1 ) )} ];

TransformPrinter : 
T0_g[ bS18{( 1 * 1 )} ]
 root domain : (bS0{1},bS1{1})
  Merge: bS0{1} and bS1{1} -> bS18{( 1 * 1 )}
T2_l[ bS17{( 1 * 1 )} ]
 root domain : (bS5{1},bS6{1})
  Merge: bS5{1} and bS6{1} -> bS17{( 1 * 1 )}
T3_l[ bS16{( 1 * ( 1 * 1 ) )} ]
 root domain : (bS7{1},bS8{1},bS9{1})
  Merge: bS8{1} and bS9{1} -> bS15{( 1 * 1 )}
  Merge: bS7{1} and bS15{( 1 * 1 )} -> bS16{( 1 * ( 1 * 1 ) )}
T1_g[ iS20{( i4 * ( 1 * 1 ) )} ]
 root domain : (iS2{i4},bS3{1},bS4{1})
  Merge: bS3{1} and bS4{1} -> bS19{( 1 * 1 )}
  Merge: iS2{i4} and bS19{( 1 * 1 )} -> iS20{( i4 * ( 1 * 1 ) )}
T4_g[ iS14{( i4 * ( 1 * 1 ) )} ]
 root domain : (iS10{i4},bS11{1},bS12{1})
  Merge: bS11{1} and bS12{1} -> bS13{( 1 * 1 )}
  Merge: iS10{i4} and bS13{( 1 * 1 )} -> iS14{( i4 * ( 1 * 1 ) )}
}
```

In this example `bS17` is `exact` mapped to `bS13` but is not permissive mapped to `bS13`, which was the part that confused the indexing pass. 

Semantically, depending on how we want to define permissive mapping, it might make sense to say all of `bS13, iS10, and iS14` permissively mapped to `bS17`, but we might lose precision and consistency quite quickly once we start to map iterdomains on the same tensordomain to each other. This would be the formalization part that'd need some follow up work.

### CODE REVIEW

**Review by csarofeen — APPROVED** [review]
WAR LGTM, maybe we should file an issue to track the questions on formalization of permissive mapping.

File: torch/csrc/jit/codegen/cuda/lower_index_compute.cpp [hunk 1]
```diff
@@ -732,8 +732,10 @@ void LoopIndexingAnalysis::constructLoopDomains() {
               !concrete_id_to_consumer_.count(concrete_id) &&
               // Use permissive map so the selected ID indeed represents the
               // loop.
-              GpuLower::current()->caMap()->areMapped(
-                  concrete_id, loop_id, IdMappingMode::PERMISSIVE);
+              (GpuLower::current()->caMap()->areMapped(
+                   concrete_id, loop_id, IdMappingMode::EXACT) ||
+               GpuLower::current()->caMap()->areMapped(
+                   concrete_id, loop_id, IdMappingMode::PERMISSIVE));
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_index_compute.cpp:764]: Can you please leave a comment here about why both maps need to be consulted?

**Review by naoyam — APPROVED** [review]
LGTM as well

File: torch/csrc/jit/codegen/cuda/lower_index_compute.cpp (hunk 1 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_index_compute.cpp:764]: updated. Thanks.

**Pull request merged**